src/main/java/org/bytedeco/javacpp/tools/Generator.java

```diff
@@ -181,6 +183,124 @@ public Generator(Logger logger, Properties properties, String encoding) {
+	out.println("");
+	out.println("JavaCPP_JavaCPP_downcast_cache_t JavaCPP_downcast_cache;");
+	out.println("");
+	out.println("extern \"C++\" template<class BASE> jobject JavaCPP_downcast(JNIEnv* env, BASE* obj) {");
+	out.println("    std::type_index id = typeid(*obj);");
+	out.println("    JavaCPP_JavaCPP_downcast_cache_t::iterator ctor = JavaCPP_downcast_cache.find(id);");
+	out.println("    if (ctor != JavaCPP_downcast_cache.end()) {");
+	out.println("        jobject ret = (jobject)env->NewObject(ctor->second.clazz, ctor->second.ctor_mid, NULL);");
+	out.println("        if (ret == NULL || env->ExceptionCheck()) {");
+	out.println("            std::cerr <<  \"Error calling Pointer constructor \" << std::endl;");
+	out.println("            return NULL;");
+	out.println("            }");
+	out.println("        if (JAVACPP_DEBUG_DCAST) std::cerr << \"Success! \" << std::endl;");
+	out.println("        return ret;");
+	out.println("        }");
+	out.println("    else");
+	out.println("        return NULL;");
+	out.println("}");
+	out.println("");
+	out.println("void JavaCPP_dcast_preload(JNIEnv* env, std::string class_name, std::type_index type_id) {");
+	out.println("    if (JAVACPP_DEBUG_DCAST) std::cerr << \"Try to find class \" <<  class_name << std::endl;");
+	out.println("");
+	out.println("    jclass lclazz = env->FindClass(class_name.c_str());");
+	out.println("    if (lclazz == NULL || env->ExceptionCheck()) {");
+	out.println("        std::cerr << \"Error loading class \" <<  class_name.c_str() << std::endl;");
+	out.println("        return;");
+	out.println("        }");
+	out.println("");
+	out.println("    jclass gclazz = (jclass)env->NewGlobalRef(lclazz);");
+	out.println("    if (gclazz == NULL || env->ExceptionCheck()) {");
+	out.println("        std::cerr << \"Error creating class \" <<  class_name  << std::endl;");
+	out.println("        return;");
+	out.println("        }");
+	out.println(" ");
+	out.println("    if (JAVACPP_DEBUG_DCAST) std::cerr << \"Try to call constructor for class \" <<  class_name << std::endl;");
+	out.println("    jmethodID ctor_mid = env->GetMethodID(gclazz, \"<init>\", \"(Lorg/bytedeco/javacpp/Pointer;)V\");");
+	out.println("    if (ctor_mid == NULL || env->ExceptionCheck()) {");
+	out.println("        std::cerr <<  \"Error getting Pointer constructor \" <<  class_name  << std::endl;");
+	out.println("        return;");
+	out.println("        }");
+	out.println("    JavaCPP_downcast_cache[type_id] = constructor{gclazz,ctor_mid};");
+	out.println("    }");
+}
+
+void outputDowncastCache(LinkedHashSet<Class> allClasses) {
+	LinkedHashSet<String> baseClasses = new LinkedHashSet<String>();
+	// Scan to find out Downcast base classes
+	for (Class<?> cls : allClasses) {
+		// if super class is InfoMapper then this is likely a target class else skip this
+		if (!InfoMapper.class.isAssignableFrom(cls.getSuperclass()))
+			continue;
```

Comment:
Let's also make this work for classes written manually that are not meant to parse header files, that is to say, they do not implement InfoMapper.

Reply:
Ok, how do we go about that?

Reply:
I suppose if we put annotations only on the base classes we're not going to need this logic, am I right?